### ISSUE TITLE
`let` nodes don't play well with `beforeEach` blocks

### ISSUE THREAD

**Issue by supermarin** [kiwi-bdd/Kiwi#588]
The following example demonstrates a bug with `let` nodes and `beforeEach` blocks.

**Expected:**  `appInfo.constants` should be initialized
**Actual:**  `appInfo.constants` is `null`

``` objc

#import <Kiwi/Kiwi.h>


SPEC_BEGIN(XYZAbandonmentSpec)

describe(@"NUX Abandonment", ^{

    let(appInfo,     ^AppInfo *{ return [AppInfo new]; });
    let(user,        ^User    *{ return [User new]; });
    let(abandonment, ^XYZ     *{ return [XYZAbandonment new]; });

    beforeEach(^{
        [[EvilSingleton sharedSingleton] stub:@selector(appInfo) andReturn:appInfo];
        [[EvilSingleton sharedSingleton] stub:@selector(user) andReturn:user];

        appInfo.constants = [AppInfoConstant new];
    });

    it(@"activates abandonment when a secret sauce", ^{
        [[@([abandonment isActive]) should] beTrue];
    });

    it(@"deactivates abandonment without secret sauce", ^{
        user.CouponAvailable = @(NO);
        [[@([abandonment isActive]) should] beFalse];
    });

    context(@"labels", ^{

        it(@"knows what the notification should say", ^{
            appInfo.constants.regLabel = @"BWOKEN!";
            NSLog(@"CONSTANTS: %@", appInfo.constants); // (null)
            [[abandonment.regLabel should] equal:@"BWOKEN!"];
        });

        it(@"knows the coupon code", ^{
            appInfo.constants.CouponCode = @"MARIN123!";
            NSLog(@"CONSTANTS: %@", appInfo.constants); // (null)
            [[abandonment.CouponCode should] equal:@"MARIN123!"];
        });
    });


});

SPEC_END
```

**Comment by bilby91** [comment]
+1

**Comment by otaran** [comment]
+1

**Comment by maurobender** [comment]
+1

**Comment by GregoryMaks** [comment]
+1 :(

**Comment by orta** [comment]
Would any of you like to try fixing it instead of a plus 1?

This is an objective-C codebase with tests, you should be fine to contribute back.

**Comment by GregoryMaks** [comment]
@orta yep, already working on it. Just discovered the mentioned issue and it took me some time to find out the reason.

**Comment by orta** [comment]
Awesomebeans

**Comment by GregoryMaks** [comment]
I made some research and tried to provide a reasonable fix, but found out that my fix breaks some tests that seems to have some sense in them. Let me explain.

If you look at the test below (from `KWFunctionalTests.m : 140`) you may see how the logic goes here. 
First of all first two 'lets' are computed, next we verify the variables, they are fine, after this we go inside the first context and override the 'subject' variable.

    describe(@"Greeting", ^{ {
        let(subject, ^{ return @""; });
        let(greeting, ^{ return [NSString stringWithFormat:@"Hello, %@!", subject]; });

        describe(@"default subject", ^{
            specify(^{ [[subject should] beEmpty]; });
        });

        context(@"with the subject \"world\"", ^{
            let(subject, ^{ return @"world"; });

            specify(^{ [[greeting should] equal:@"Hello, world!"]; });
        });

        context(@"with the subject \"Kiwi\"", ^{
            let(subject, ^{ return @"Kiwi"; });

            specify(^{ [[greeting should] equal:@"Hello, Kiwi!"]; });
        });
    });

What would you expect 'greeting' here to be? I expected it to be calculated only once for each test, thus it makes sense that it equals "Hello,  !".

But in current implementation every let variable outside the scope is recalculated once you go inside the scope.
This is what makes it immediately equal to "Hello, world!". And that is why the issue mentioned above is happening.

In @supermarin case the sequence of calls goes as following:

    1. let(appInfo,     ^AppInfo *{ return [AppInfo new]; });

    2. beforeEach(^{
        ...
        appInfo.constants = [AppInfoConstant new];
    });

    3. [we go inside context] context(@"labels", ^{

    4. [recalculate all vars, CONSTANTS ARE LOST!] let(appInfo,     ^AppInfo *{ return [AppInfo new]; });

    5. [go inside context's inner test] it(@"knows what the notification should say", ^{

    6. [all changes made to 'appInfo' in 'beforeEach' are logically lost]


I have prepared a fix and and a test to change the behavior, but as it breaks some other tests I would like to have some input from the creators. 

_What way of using variables here should be the right one?_

P.S.: I stand on the idea of 'calculate variables only once for each test no matter how deep inside any scope we go' because variables not lazily computed here, nor are they dynamic in any way. Mixing ideology of normal variable and dynamic variable is confusing people.

**Comment by orta** [comment]
So this is only my opinion, and it only come from using it in Rspec - but the idea of dynamic scope based re-declarations of `let` feels off to me. 

IMO this should raise:

```
describe(@"Greeting", ^{ {
    let(subject, ^{ return @""; });

    context(@"with the subject \"world\"", ^{
        let(subject, ^{ return @"world"; }); // <- here
        specify(^{ [[greeting should] equal:@"Hello, world!"]; });
    });
```

Related to that, Step 4 feels like a bug to me, if the let's are being re-evaluated on a per-describe context. @supermarin's example seems like a much more common use case to me
